sandbox.go

```diff
@@ -1,130 +1,317 @@
+	// 	context.TODO(),
+	// 	networkName,
+	// 	types.NetworkCreate{},
+	// )
+	// if err != nil {
+	// 	return nil, err
+	// }
+
+	s = &sandbox{
+		ID:       uuid.NewV4().String(),
+		language: l,
+		code:     code,
+		stdin:    stdin,
+		options:  opts,
+		docker:   dockerClient,
+	}
+
+	return s, nil
 }
 
+// runs user code within the sandbox after preparing the execution environment.
+//
 func (s *sandbox) run() (string, error) {
-	s.prepare()
-	return s.execute()
+	var (
+		output string
+		err    error
+	)
+
+	err = s.prepare()
+	if err != nil {
+		return "", err
+	}
+
+	output, err = s.execute()
+	if err != nil {
+		return "", err
+	}
+
+	return output, nil
 }
 
-func (s *sandbox) prepare() {
-	tmpFolder, err := ioutil.TempDir(s.options.folder, "xaqt-")
+// prepares the execution environment and the sandbox docker container.
+//
+func (s *sandbox) prepare() error {
+	var err error
+
+	err = s.PrepareTmpDir()
+	if err != nil {
+		return err
+	}
+
+	err = s.PrepareContainer()
 	if err != nil {
-		log.Fatal(err)
+		return err
+	}
+
+	return nil
+}
+
+// prepares the execution environment by copying all resources (user code, input files,
+// and execution payload) into a temporary directory.
+//
+func (s *sandbox) PrepareTmpDir() error {
+	// create tmp directory for keeping all code and inputs
+	tmpFolder, err := ioutil.TempDir(s.options.folder, TmpDirPrefix)
+	if err != nil {
+		return err
 	}
 
+	// modify perms on tmp dir
 	if err := os.Chmod(tmpFolder, 0777); err != nil {
-		log.Fatal(err)
+		return err
 	}
 
 	// record tmpdir for easy deletion
 	s.options.folder = tmpFolder
 
+	// copy the Payload dir into the tmp dir. for more details on what the
+	// Payload dir is, check out the TODO (cw|4.29.2018) fill this in...
 	err = s.copyPayload()
 	if err != nil {
-		log.Fatal(err)
+		return err
 	}
 
-	//log.Println(s.code)
+	// write source file into tmp dir
+	// TODO (cw|4.29.2018) we should be able to write an arbitrary number of files
+	// to the tmp dir.
 	err = ioutil.WriteFile(tmpFolder+"/"+s.language.SourceFile, []byte(s.code), 0777)
 	if err != nil {
-		log.Fatal(err)
+		return err
 	}
 
-	//fmt.Println(ioutil.ReadFile(tmpFolder + "/" + s.language.SourceFile))
-
 	// write a file for stdin
-	log.Printf("writing inputfile: %v", s.stdin)
 	err = ioutil.WriteFile(tmpFolder+"/inputFile", []byte(s.stdin), 0777)
 	if err != nil {
-		log.Fatal(err)
+		return err
 	}
-	// log msg
+
+	return nil
 }
 
-func (s *sandbox) execute() (string, error) {
-	defer os.RemoveAll(s.options.folder)
+// create docker container for running code and stream container's stdout to our stdout.
+//
+func (s *sandbox) PrepareContainer() error {
+	var (
+		ctx = context.Background()
+		err error
+	)
 
-	compiler := s.language.Compiler
-	filename := s.language.SourceFile
-	optionalExecutable := s.language.OptionalExecutable
-	flags := s.language.CompilerFlags
+	// create docker container for executing user code
+	_, err = s.docker.ContainerCreate(
```

Comment:
Do we need to close/destroy/cleanup this container in some fashion within a `defer` block?

Reply:
when the container is configured it has been set to be auto removed once the process in the container completes (see [this part of the container config](https://github.com/frenata/xaqt/pull/21/files/7f3b56bfe511b62be34d23a31f72c2e583dbdb7f#diff-ffc9c06e306285f214919ed203907012R206))